### PR TITLE
ARROW-17016: [C++][Python] Move Arrow Python C++ tests into Cython

### PR DESCRIPTION
**Pull request by AlenkaF**
This PR tries to connect the PyArrow C++ tests with PyArrow tests so they can all be run from `pytest`. This will remove GoogleTest as a dependency for PyArrow and therefore a change is needed in the C++ tests so they return a Status which can then be checked through Cython/Python.

Example of pytest run:

```
pyarrow/tests/test_cpp_internals.py::test_owned_ref_moves PASSED
pyarrow/tests/test_cpp_internals.py::test_owned_ref_nogil_moves PASSED
pyarrow/tests/test_cpp_internals.py::test_check_pyerror_status PASSED
pyarrow/tests/test_cpp_internals.py::test_check_pyerror_status_nogil PASSED
pyarrow/tests/test_cpp_internals.py::test_restore_pyerror_basics PASSED
pyarrow/tests/test_cpp_internals.py::test_pybuffer_invalid_input_object PASSED
pyarrow/tests/test_cpp_internals.py::test_pybuffer_numpy_array PASSED
pyarrow/tests/test_cpp_internals.py::test_numpybuffer_numpy_array PASSED
pyarrow/tests/test_cpp_internals.py::test_python_decimal_to_string PASSED
pyarrow/tests/test_cpp_internals.py::test_infer_precision_and_scale PASSED
pyarrow/tests/test_cpp_internals.py::test_infer_precision_and_negative_scale PASSED
pyarrow/tests/test_cpp_internals.py::test_infer_all_leading_zeros PASSED
pyarrow/tests/test_cpp_internals.py::test_infer_all_leading_zeros_exponential_notation_positive PASSED
pyarrow/tests/test_cpp_internals.py::test_infer_all_leading_zeros_exponential_notation_negative PASSED
pyarrow/tests/test_cpp_internals.py::test_object_block_write_fails PASSED
pyarrow/tests/test_cpp_internals.py::test_mixed_type_fails PASSED
pyarrow/tests/test_cpp_internals.py::test_from_python_decimal_rescale_not_truncateable PASSED
pyarrow/tests/test_cpp_internals.py::test_from_python_decimal_rescale_truncateable PASSED
pyarrow/tests/test_cpp_internals.py::test_from_python_negative_decimal_rescale PASSED
pyarrow/tests/test_cpp_internals.py::test_decimal128_from_python_integer PASSED
pyarrow/tests/test_cpp_internals.py::test_decimal256_from_python_integer PASSED
pyarrow/tests/test_cpp_internals.py::test_decimal128_overflow_fails PASSED
pyarrow/tests/test_cpp_internals.py::test_decimal256_overflow_fails PASSED
pyarrow/tests/test_cpp_internals.py::test_none_and_nan PASSED
pyarrow/tests/test_cpp_internals.py::test_mixed_precision_and_scale PASSED
pyarrow/tests/test_cpp_internals.py::test_mixed_precision_and_scale_sequence_convert PASSED
pyarrow/tests/test_cpp_internals.py::test_simple_inference PASSED
pyarrow/tests/test_cpp_internals.py::test_update_with_nan PASSED
```

### CODE REVIEW

File: python/pyarrow/src/python_test.cc [hunk 1]
```diff
@@ -92,81 +131,115 @@ std::string FormatPythonException(const std::string& exc_class_name) {
...
+  //     auto detail = st.detail();
+  //     ASSERT_NE(detail, nullptr);
+  //     ASSERT_EQ(detail->ToString(), expected_detail);
+  //   }
+  // };
 
   for (PyObject* exc_type : {PyExc_Exception, PyExc_SyntaxError}) {
     PyErr_SetString(exc_type, "some error");
-    check_error(st);
+    //check_error(st);
```
[the remark this answers is not in the document]
Comment (AlenkaF) [reply python/pyarrow/src/python_test.cc:152]: Aha, that I tried but was missing `ASSERT_OK(check_error(st));`. Thank you so much!!

File: python/setup.cfg [hunk 2]
```diff
@@ -28,6 +28,9 @@ build-dir  = doc/_build
 addopts = --ignore=scripts
 filterwarnings =
     error:The SparseDataFrame:FutureWarning
+    # ignore warnings when Cython functions are called
+    # from python tests
+    ignore::pytest.PytestCollectionWarning
```
Comment (pitrou) [inline python/setup.cfg:33]: What is the warning that this would silence?

**Review by pitrou — COMMENTED** [review]
Just some general comments. I will probably want to make some slight changes on the C++ side as well.

File: python/setup.cfg (hunk 2 above)
Comment (AlenkaF) [reply python/setup.cfg:33]: Oh, this is a leftover from when I wanted to make Cython tests to be picked up by Pytest automatically. Didn't work, need to delete it!

File: python/pyarrow/_pyarrow_cpp_tests.pxd [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+cdef extern from "arrow/python/python_test.h" namespace "arrow::py" nogil:
+
+    CStatus TestOwnedRefMoves()
+    CStatus TestOwnedRefNoGILMoves()
+    CStatus TestCheckPyErrorStatus()
+    CStatus TestCheckPyErrorStatusNoGIL()
+    CStatus TestRestorePyErrorBasics()
+    CStatus TestPyBufferInvalidInputObject()
+
+    # ifdef _WIN32
```
[the remark this answers is not in the document]
Comment (AlenkaF) [reply python/pyarrow/_pyarrow_cpp_tests.pxd:33]: Well, it was a try&error kind of thing - AppVeyor doesn't look happy about it.

File: python/pyarrow/tests/test_python.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_python.py:1]: Small nit, but maybe call this `test_cpp.py` or `test_pyarrow_cpp.py` instead? ("python" is a bit generic)

File: python/pyarrow/_pyarrow_cpp_tests.pxd (hunk 3 above)
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply python/pyarrow/_pyarrow_cpp_tests.pxd]: Not relevant anymore for the current PR, but cython gives some way to use such `#if`'s, but if you put those in the "docstring" of this with block: https://docs.cython.org/en/latest/src/userguide/external_C_code.html#including-verbatim-c-code

File: python/pyarrow/tests/test_python.py (hunk 4 above)
Comment (pitrou) [reply python/pyarrow/tests/test_python.py:1]: +1, or also `test_cpp_internals.py`

Comment (jorisvandenbossche) [reply python/pyarrow/tests/test_python.py:1]: Yes, that sounds good

Comment (AlenkaF) [reply python/pyarrow/tests/test_python.py:1]: OK, will do 👍

**Comment by AlenkaF** [comment]
Should we wait for [ARROW-17843](https://issues.apache.org/jira/browse/ARROW-17843) so that the PyArrow C++ code is checked or we can correct that within the linter issue itself?

**Comment by pitrou** [comment]
We can correct the C++ linting later IMHO.

**Comment by jorisvandenbossche** [comment]
@AlenkaF can you merge in or rebase on latest master? That should resolve those CI failures

**Review by pitrou — APPROVED** [review]
+1, thank you very much for doing this @AlenkaF !

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python -g wheel

**Comment by AlenkaF** [comment]
I think the failures of the wheels are related to the fact that `ARROW_PYTHON` is not set anymore and so for example csv module is not build and not found by PyArrow (`ARROW_CSV=OFF`).

**Comment by pitrou** [comment]
Yes, we will probably need to enable all desired components manually.

**Comment by jorisvandenbossche** [comment]
Yes, that is getting fixed in https://github.com/apache/arrow/pull/14273

**Comment by AlenkaF** [comment]
You think I should wait for https://github.com/apache/arrow/pull/14273 to get merged and then rebase?

**Comment by AlenkaF** [comment]
@pitrou @jorisvandenbossche this PR is now ready for final round of review.

**Comment by jorisvandenbossche** [comment]
@github-actions crossbow submit -g python -g wheel

**Comment by AlenkaF** [comment]
Thank you for running the wheels build - totally forgot that was what I was waiting to check :)

**Pull request merged**